Adding Module that Utilizes Azure Key Vault (#1)
nexilus18/kube-workshop: 12-using-az-kv/readme.md

# Using Azure Key Vault to store secrets

> 

Reviewer: Introduction section on Azure Key Vault approach vs AKS Secrets described in prior lab

## Set up

- Install latest version of the `aks-preview` extension or update to the latest version
  ```bash
  az extension add --name aks-preview
  or
  az extension update --name aks-preview
  ```

> - Existing Azure Subscription with EnableWorkloadIdentityPreview feature enabled

Reviewer: ```suggestion
- Enable `EnableWorkloadIdentityPreview` feature on the existing Azure Subscription 
```

Reviewer: It would be helpful to add the command to add this feature or link out to the page

Update vars.sh file to include few more variables for using Azure Key Vault

```yaml
SUBSCRIPTION_ID="__change_me__" #subscription id
UAMI="__change_me__"     #name for user assigned identity
KEYVAULT_NAME="__change_me__" #keyvault name
IDENTITY_TENANT="__change_me__" #tenant id
```

Source the file through with `source vars.sh`.

## Upgrade existing AKS cluster

> Upgrade your AKS cluster with Azure Key Vaut Provider for Secrets store CSI Driver capability, enable-oidc-issuer

Reviewer: ```suggestion
Upgrade your AKS cluster with Azure Key Vault Provider for Secrets store CSI Driver capability, `enable-oidc-issuer`
```

```bash
az aks enable-addons --addons azure-keyvault-secrets-provider --name $AKS_NAME --resource-group $RES_GROUP
az aks update -g $RES_GROUP -n $AKS_NAME --enable-oidc-issuer

```

## Create Azure KeyVault

> Create an Azure Key Vault resource to store secrets. You can store secret in your key vault using the `az keyvaul secret set` command

Reviewer: ```suggestion
Create an Azure Key Vault resource to store secrets. You can store the key-value pair of `admin-password` and related secret in the key vault using the `az keyvault secret set` command. The `admin-password` is the name of the secret and the actual secret is stored in value.
```

```bash
az keyvault create -n $KEYVAULT_NAME -g $RES_GROUP -l $REGION
```

```bash

> az keyvault secret set --vault-name $KEYVAULT_NAME -n admin-passsword --value kv_supersecret

Reviewer: ```suggestion
az keyvault secret set --vault-name $KEYVAULT_NAME -n admin-passsword --value supersecret
```

```

## Configure Workload Identity

> First set a specific subscription to be the current active subscription:

Reviewer: ```suggestion
First, set a specific subscription to be the current active subscription:
```

```bash
az account set --subscription $SUBSCRIPTION_ID
```

Create managed identity using `az identity create`:

```bash
az identity create --name $UAMI --resource-group $RES_GROUP

```

> The following command outputs the user assigned client id. Save this output

Reviewer: auto add to vars.sh file

in the `vars.sh` as USER_ASSIGNED_CLIENT_ID.

```bash
az identity show -g $RES_GROUP --name $UAMI --query 'clientId' -o tsv
```

In `vars.sh`, set the following variable:

```yaml
USER_ASSIGNED_CLIENT_ID="__change_me__"
```

Run `source vars.sh` before moving the next step.

You need to grant the workload identity permission to access the Key Vault
secrets, keys and certificates by setting acesss policy.

```bash
az keyvault set-policy -n $KEYVAULT_NAME --key-permissions get --spn $USER_ASSIGNED_CLIENT_ID
az keyvault set-policy -n $KEYVAULT_NAME --secret-permissions get --spn $USER_ASSIGNED_CLIENT_ID
az keyvault set-policy -n $KEYVAULT_NAME --certificate-permissions get --spn $USER_ASSIGNED_CLIENT_ID
```

```bash
export AKS_OIDC_ISSUER="$(az aks show --resource-group $RES_GROUP --name $AKS_NAME --query "oidcIssuerProfile.issuerUrl" -o tsv)"
echo $AKS_OIDC_ISSUER
```

_If URL is empty, verify you have installed the latest version of the `aks-preview` extension andd OIDC issuer is enabled._

Update the values for `name` and `namespace` with the Kubernetes service account name and its namespace in the `service.yaml` file and apply the manifest.

<details markdown="1">

> <summary>Click here for the Service YAML file</summary>

Reviewer: ```suggestion
<summary>Click here for the Service.yaml file</summary>
```

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  annotations:
    azure.workload.identity/client-id: ${USER_ASSIGNED_CLIENT_ID}
  labels:
    azure.workload.identity/use: "true"
  name: "__change_me__" # service account name
  namespace: "_change_me__" # namespace of your workload
```

</details>

```bash
kubectl apply -f service.yaml
```

> Create the federated identity credential between the Managed Identity, the service account issuer, and the subject.

Reviewer: Note: to add link to federated identity credential

```bash
export federatedIdentityName="aksfederatedidentity" # can be changed as needed
az identity federated-credential create --name $federatedIdentityName --identity-name $UAMI --resource-group $RES_GROUP --issuer ${AKS_OIDC_ISSUER} --subject system:serviceaccount:${serviceAccountNamespace}:${serviceAccountName}
```

> In the secret-provider-class, change `ClientID`, `keyvaultName` and `tenantID` with the worload identity, keyvault name and the tenant ID of the keyvault. Update `ObjectName` with the name of your secret. Deploy the `SecretProviderClass` after making the changes.

Reviewer: ```suggestion
In the secret-provider-class, change `ClientID`, `keyvaultName` and `tenantID` with the workload identity, keyvault name and the tenant ID of the keyvault. Update `ObjectName` with the name of your secret. Deploy the `SecretProviderClass` after making the changes.
```

<details markdown="2">

> <summary>Click here for the Secret Provider Class YAML</summary>

Reviewer: .yaml file
